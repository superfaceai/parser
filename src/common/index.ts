export * from './document/interfaces';
export * from './document/parser';
export * from './split';
