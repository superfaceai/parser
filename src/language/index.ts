export { Location, Span, Source } from './source';
export { SyntaxError } from './error';

export * from './lexer';
export * from './syntax';
