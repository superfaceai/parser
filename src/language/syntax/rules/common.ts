import {
  DocumentedNode,
  MapASTNodeBase,
  ProfileASTNodeBase,
} from '@superfaceai/language';

import { SyntaxRule } from '../rule';
import { extractDocumentation } from '../util';

// HELPER RULES //

type ASTNodeBase = ProfileASTNodeBase | MapASTNodeBase;

// Node that has `span` and `location` non-optional.
export type SrcNode<N extends ASTNodeBase> = N & {
  span: NonNullable<N['span']>;
  location: NonNullable<N['location']>;
};
export type SyntaxRuleSrc<N extends ASTNodeBase> = SyntaxRule<SrcNode<N>>;

export function documentedNode<
  N extends SrcNode<DocumentedNode & ASTNodeBase>,
  R extends SyntaxRule<N>
>(rule: R): SyntaxRule<N> {
  return SyntaxRule.optional(SyntaxRule.string())
    .followedBy(rule)
    .map(
      (matches): N => {
        const [maybeDoc, result] = matches;
        if (maybeDoc !== undefined) {
          const doc = extractDocumentation(maybeDoc.data.string);
          result.title = doc.title;
          result.description = doc.description;
          result.location = maybeDoc.location;
          result.span.start = maybeDoc.span.start;
        }

        return result;
      }
    );
}
