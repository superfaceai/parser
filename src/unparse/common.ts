import { DocumentedNode } from '@superfaceai/ast';

export abstract class UnparserBase {
  protected currentDepth: number;
  protected readonly options: { indent: string };

  constructor(options?: { indent: string }) {
    this.currentDepth = 0;
    this.options = {
      indent: '  ',
      ...options,
    };
  }

  protected indentJoinLines(...strings: (string | undefined)[]): string {
    let result = '';

    for (const str of strings) {
      if (str !== undefined) {
        result +=
          this.options.indent.repeat(this.currentDepth) + str.trimLeft() + '\n';
      }
    }

    return result;
  }

  protected stripLast(str: string): string {
    return str.slice(0, str.length - 1);
  }

  protected visitDocumentedNode(node: DocumentedNode): string | undefined {
    let doc = node.documentation?.title;
    if (node.documentation?.description !== undefined) {
      doc = (doc ?? '') + '\n\n' + node.documentation?.description;
    }

    if (doc !== undefined) {
      doc = `"""\n${doc}\n"""`;
      doc = this.indentJoinLines(...doc.split('\n'));
    }

    return doc;
  }
}
