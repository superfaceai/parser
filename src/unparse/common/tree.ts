export type UnparserTokenKind =
  | 'Atom'
  | 'Space'
  | 'Newline'
  | 'Block'
  | 'Condition';
export type UnparserTokenBase = {
  kind: UnparserTokenKind;
};

export type AtomToken = UnparserTokenBase & {
  kind: 'Atom';
  value: string;
};
export type SpaceToken = UnparserTokenBase & {
  kind: 'Space';
  count: number;
};
export type NewlineToken = UnparserTokenBase & {
  kind: 'Newline';
};
export type BlockToken = UnparserTokenBase & {
  kind: 'Block';
  symbols: '{}' | '()' | '[]' | '<>';
  tokens: UnparserToken[];
};
export type ConditionToken = UnparserTokenBase & {
  kind: 'Condition';
  /** Required configuration to enable materialization of this token. */
  condition: string;
  positive: UnparserTokenGroup;
  negative: UnparserTokenGroup;
};

export type UnparserToken =
  | AtomToken
  | SpaceToken
  | NewlineToken
  | BlockToken
  | ConditionToken;
export type UnparserTokenGroup = UnparserToken[];

// UnparserToken factory
export const UnparserToken = {
  /** Produces a token group glued together by spaces.
   *
   * Strings are turned into atoms. Undefined elements are filtered out.
   */
  withSpaces(
    ...values: (UnparserToken | UnparserTokenGroup | string | undefined)[]
  ): UnparserTokenGroup {
    return UnparserToken.addGlue(
      UnparserToken.space(),
      values.map(v => (typeof v === 'string' ? UnparserToken.atom(v) : v))
    );
  },

  if(
    condition: ConditionToken['condition'],
    positive: UnparserTokenGroup | UnparserToken,
    negative?: UnparserTokenGroup | UnparserToken
  ): ConditionToken {
    const neg = negative ?? [];

    return {
      kind: 'Condition',
      condition,
      positive: Array.isArray(positive) ? positive : [positive],
      negative: Array.isArray(neg) ? neg : [neg],
    };
  },

  /** Adds `glue` between tokens or token groups.
   *
   * Undefined elements are filtered out.
   */
  addGlue(
    glue: UnparserToken | UnparserTokenGroup,
    tokens: (UnparserToken | UnparserTokenGroup | undefined)[]
  ): UnparserToken[] {
    const notUndefined = (
      v: UnparserToken | UnparserTokenGroup | undefined
    ): v is UnparserToken | UnparserTokenGroup => v !== undefined;

    const glueGroup = Array.isArray(glue) ? glue : [glue];

    return tokens
      .filter(notUndefined)
      .flatMap<UnparserToken>((tokenGroup, index, arr) => {
        if (index < arr.length - 1) {
          if (Array.isArray(tokenGroup)) {
            return [...tokenGroup, ...glueGroup];
          }

          return [tokenGroup, ...glueGroup];
        }

        return tokenGroup;
      });
  },

  atom(value: string): AtomToken {
    return { kind: 'Atom', value };
  },
  space(count?: number): SpaceToken {
    return { kind: 'Space', count: count ?? 1 };
  },
  newline(): NewlineToken {
    return { kind: 'Newline' };
  },

  genericBlock(
    symbols: BlockToken['symbols'],
    lead: UnparserTokenGroup,
    glue: UnparserTokenGroup,
    trail: UnparserTokenGroup,
    items: UnparserTokenGroup[]
  ): BlockToken {
    return {
      kind: 'Block',
      symbols,
      tokens: [
        ...lead,
        ...items.flatMap<UnparserToken>((itemLine, index, arr) => {
          if (index < arr.length - 1) {
            return [...itemLine, ...glue];
          } else {
            return itemLine;
          }
        }),
        ...trail,
      ],
    };
  },
};

export type UnparserTreeMaterializerConfiguration = {
  identString: string;
  conditionValues: Record<string, boolean>;
};
export class UnparserTreeMaterializer {
  private result = '';
  private depth = 0;
  private lineState: 'start' | 'continue' = 'start';
  /** Number of consecutive spaces since the last token. */
  private spaceCount = 0;

  constructor(
    private readonly configuration: UnparserTreeMaterializerConfiguration
  ) {}

  private resolveCondition(token: ConditionToken): UnparserTokenGroup {
    if (!(token.condition in this.configuration.conditionValues)) {
      throw new Error('UnparserTreeMaterializer: Invalid condition key');
    }

    let result = token.negative;
    if (this.configuration.conditionValues[token.condition]) {
      result = token.positive;
    }

    return result;
  }

  materialize(tree: UnparserToken[]): string {
    // reset all state
    this.result = '';
    this.depth = 0;
    this.lineState = 'start';
    this.spaceCount = 0;

    for (const token of tree) {
      this.materializeRecursive(token);
    }

    return this.result;
  }

  private emit(s: string) {
    switch (this.lineState) {
      case 'start':
        this.lineState = 'continue';
        this.result += this.configuration.identString.repeat(this.depth) + s;
        break;

      case 'continue':
        this.result += s;
        break;
    }
  }

  private materializeRecursive(token: UnparserToken) {
    switch (token.kind) {
      case 'Atom':
        this.emit(token.value);
        this.spaceCount = 0;
        break;

      case 'Space':
        if (this.spaceCount < token.count) {
          this.emit(' '.repeat(token.count - this.spaceCount));
          this.spaceCount = token.count;
        }
        break;

      case 'Newline':
        this.lineState = 'start';
        this.result += '\n';
        this.spaceCount = 0;
        break;

      case 'Block':
        this.emit(token.symbols[0]);
        this.spaceCount = 0;

        this.depth += 1;
        for (const subtoken of token.tokens) {
          this.materializeRecursive(subtoken);
        }
        this.depth -= 1;

        this.emit(token.symbols[1]);
        this.spaceCount = 0;
        break;

      case 'Condition':
        for (const subtoken of this.resolveCondition(token)) {
          this.materializeRecursive(subtoken);
        }
        break;
    }
  }
}
