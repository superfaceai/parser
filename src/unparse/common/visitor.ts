import {
  UnparserTokenGroup,
  UnparserTreeMaterializer,
  UnparserTreeMaterializerConfiguration,
} from './tree';

export abstract class UnparserBase<T> {
  private readonly materializer: UnparserTreeMaterializer;

  constructor(configuration: UnparserTreeMaterializerConfiguration) {
    this.materializer = new UnparserTreeMaterializer(configuration);
  }

  unparse(ast: T): string {
    const tree = this.visit(ast);

    return this.materializer.materialize(tree);
  }

  abstract visit(ast: T): UnparserTokenGroup;
}
