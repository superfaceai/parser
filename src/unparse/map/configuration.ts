import { UnparserTreeMaterializerConfiguration } from '../common/tree';

export type MapUnparserConfiguration = {
  /** String used for one level of indentation. */
  indentString: string;
  /** Whether to add trailing terminators (commas, semicolons). */
  trailingTerminators: boolean;
  /** Collapsed set blocks are on one line only and semicolons are mandatory. Expanded are on multiple lines with optional semicolons. */
  setBlockStyle: 'collapsed' | 'expanded' | 'expanded+semicolons';
  /** Call blocks appear in operation calls as the list of arguments between parens.
   *
   * Collapsed call blocks are on one line only and commas are mandatory. Expanded are on multiple lines with optional commas.
   */
  callBlockStyle: 'collapsed' | 'expanded' | 'expanded+commas';
  /** Collapsed object blocks are on one line only and commas are mandatory. Expanded are on multiple lines with optional commas. */
  objectBlockStyle: 'collapsed' | 'expanded' | 'expanded+commas';

  /** Whether to emit http call `default` selector. Non-default selectors are always emitted. */
  httpCallDefaultService: 'never' | 'always';
  /** Whether to emit http call `security none` selector. Non-default selectors are always emitted. */
  httpCallDefaultSecurity: 'never' | 'always';
};
export const DEFAULT_MAP_UNPARSER_CONFIGURATION: MapUnparserConfiguration = {
  indentString: '  ',
  trailingTerminators: false,
  setBlockStyle: 'expanded',
  callBlockStyle: 'collapsed',
  objectBlockStyle: 'expanded',
  httpCallDefaultService: 'never',
  httpCallDefaultSecurity: 'never',
};

export function transformMapUnparserConfiguration(
  config: MapUnparserConfiguration
): UnparserTreeMaterializerConfiguration {
  const setBlockExpand =
    config.setBlockStyle === 'expanded' ||
    config.setBlockStyle === 'expanded+semicolons';
  const setBlockSemicolons =
    config.setBlockStyle === 'collapsed' ||
    config.setBlockStyle === 'expanded+semicolons';

  const callBlockExpand =
    config.callBlockStyle === 'expanded' ||
    config.callBlockStyle === 'expanded+commas';
  const callBlockCommas =
    config.callBlockStyle === 'collapsed' ||
    config.callBlockStyle === 'expanded+commas';

  const objectBlockExpand =
    config.objectBlockStyle === 'expanded' ||
    config.objectBlockStyle === 'expanded+commas';
  const objectBlockCommas =
    config.objectBlockStyle === 'collapsed' ||
    config.objectBlockStyle === 'expanded+commas';

  return {
    identString: config.indentString,
    conditionValues: {
      setBlockNewlines: setBlockExpand,
      setBlockSpaceLeadTrail: !setBlockExpand,
      setBlockSpaceGlue: !setBlockExpand,
      setBlockSemicolonGlue: setBlockSemicolons,
      setBlockSemicolonTrail: setBlockSemicolons && config.trailingTerminators,

      callBlockNewlines: callBlockExpand,
      callBlockSpaceLeadTrail: false,
      callBlockSpaceGlue: !callBlockExpand,
      callBlockCommaGlue: callBlockCommas,
      callBlockCommaTrail: callBlockCommas && config.trailingTerminators,

      objectBlockNewlines: objectBlockExpand,
      objectBlockSpaceLeadTrail: !objectBlockExpand,
      objectBlockSpaceGlue: !objectBlockExpand,
      objectBlockCommaGlue: objectBlockCommas,
      objectBlockCommaTrail: objectBlockCommas && config.trailingTerminators,

      conditionBlockNewlines: false,
      conditionBlockSpaceLeadTrail: false,

      foreachBlockNewlines: false,
      foreachBlockSpaceLeadTrail: false,

      httpCallDefaultService: config.httpCallDefaultService === 'always',
      httpCallDefaultSecurity: config.httpCallDefaultSecurity === 'always',
    },
  };
}
