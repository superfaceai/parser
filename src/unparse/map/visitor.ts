import {
  AssignmentNode,
  CallStatementNode,
  ConditionAtomNode,
  HttpCallStatementNode,
  HttpRequestNode,
  HttpResponseHandlerNode,
  InlineCallNode,
  IterationAtomNode,
  JessieExpressionNode,
  MapASTNode,
  MapAstVisitor,
  MapDefinitionNode,
  MapDocumentNode,
  MapHeaderNode,
  ObjectLiteralNode,
  OperationDefinitionNode,
  OutcomeStatementNode,
  PrimitiveLiteralNode,
  SetStatementNode,
} from '@superfaceai/ast';

import { ProfileId, VersionRange } from '../../common';
import {
  isValidIdentifierChar,
  isValidIdentifierStartChar,
} from '../../common/source';
import { BlockToken, UnparserToken, UnparserTokenGroup } from '../common/tree';
import { UnparserBase } from '../common/visitor';
import {
  MapUnparserConfiguration,
  transformMapUnparserConfiguration,
} from './configuration';
import { UnparserTokenMap as Tok } from './tree';

export class MapUnparser
  extends UnparserBase<MapASTNode>
  implements MapAstVisitor
{
  private currentContext: 'map' | 'operation';

  constructor(configuration: MapUnparserConfiguration) {
    super(transformMapUnparserConfiguration(configuration));
    this.currentContext = 'map';
  }

  visit(node: MapASTNode): UnparserTokenGroup {
    switch (node.kind) {
      case 'MapDocument':
        return this.visitMapDocumentNode(node);
      case 'MapHeader':
        return this.visitMapHeaderNode(node);
      case 'MapDefinition':
        return this.visitMapDefinitionNode(node);
      case 'OperationDefinition':
        return this.visitOperationDefinitionNode(node);

      case 'SetStatement':
        return this.visitSetStatementNode(node);
      case 'CallStatement':
        return this.visitCallStatementNode(node);
      case 'HttpCallStatement':
        return this.visitHttpCallStatementNode(node);
      case 'OutcomeStatement':
        return this.visitOutcomeStatementNode(node);

      case 'Assignment':
        return this.visitAssignmentNode(node);
      case 'ConditionAtom':
        return this.visitConditionAtomNode(node);
      case 'IterationAtom':
        return this.visitIterationAtomNode(node);
      case 'HttpResponseHandler':
        return this.visitHttpResponseHandlerNode(node);
      case 'InlineCall':
        return this.visitInlineCallNode(node);

      case 'PrimitiveLiteral':
        return [this.visitPrimitiveLiteralNode(node)];
      case 'ObjectLiteral':
        return [this.visitObjectLiteralNode(node)];
      case 'JessieExpression':
        return [this.visitJessieExpressionNode(node)];

      default:
        throw new Error('MapUnparser: unreachable node kind ' + node.kind);
    }
  }

  protected visitCallHead(call: {
    operationName: string;
    condition?: ConditionAtomNode;
    iteration?: IterationAtomNode;
    arguments: AssignmentNode[];
  }): UnparserTokenGroup {
    return Tok.withSpaces(
      'call',
      call.iteration ? this.visit(call.iteration) : undefined,
      [
        Tok.atom(call.operationName),
        Tok.callBlock(...call.arguments.map(a => this.visit(a))),
      ],
      call.condition ? this.visit(call.condition) : undefined
    );
  }

  visitPrimitiveLiteralNode(ape: PrimitiveLiteralNode): UnparserToken {
    let value = ape.value.toString();
    if (typeof ape.value === 'string') {
      value = `"${value}"`;
    }

    return Tok.atom(value);
  }

  visitObjectLiteralNode(object: ObjectLiteralNode): BlockToken {
    return Tok.objectBlock(...object.fields.map(f => this.visit(f)));
  }

  visitJessieExpressionNode(node: JessieExpressionNode): UnparserToken {
    // TODO: formatting of inner indentation
    return Tok.atom(node.source ?? node.expression);
  }

  visitAssignmentNode(ass: AssignmentNode): UnparserTokenGroup {
    const key = ass.key
      .map(key => {
        const isValidIdentifier =
          isValidIdentifierStartChar(key.charCodeAt(0)) &&
          Array.from(key).every(ch => isValidIdentifierChar(ch.charCodeAt(0)));

        if (isValidIdentifier) {
          return key;
        } else {
          return `"${key}"`;
        }
      })
      .join('.');

    return Tok.withSpaces(key, '=', this.visit(ass.value));
  }

  visitConditionAtomNode(ana: ConditionAtomNode): UnparserTokenGroup {
    return [
      Tok.atom('if'),
      Tok.space(),
      Tok.conditionBlock(this.visit(ana.expression)),
    ];
  }

  visitIterationAtomNode(iter: IterationAtomNode): UnparserTokenGroup {
    return [
      Tok.atom('foreach'),
      Tok.space(),
      Tok.foreachBlock(
        Tok.withSpaces(iter.iterationVariable, 'of', this.visit(iter.iterable))
      ),
    ];
  }

  visitSetStatementNode(set: SetStatementNode): UnparserTokenGroup {
    if (set.assignments.length > 1 || set.condition !== undefined) {
      return Tok.withSpaces(
        'set',
        set.condition !== undefined ? this.visit(set.condition) : undefined,
        Tok.setBlock(...set.assignments.map(ass => this.visit(ass)))
      );
    } else {
      return this.visit(set.assignments[0]);
    }
  }

  visitCallStatementNode(call: CallStatementNode): UnparserTokenGroup {
    return Tok.withSpaces(
      this.visitCallHead(call),
      Tok.statementBlock(...call.statements.map(s => this.visit(s)))
    );
  }

  visitHttpRequestNode(req: HttpRequestNode): UnparserTokenGroup {
    let security: UnparserTokenGroup = [
      Tok.if('httpCallDefaultSecurity', [
        ...Tok.withSpaces('security', 'none'),
        Tok.newline(),
      ]),
    ];
    if (req.security.length > 0) {
      security = req.security.flatMap(s => [
        ...Tok.withSpaces('security', `"${s.id}"`),
        Tok.newline(),
      ]);
    }

    let contentSettings: string[] = [];
    if (req.contentLanguage !== undefined) {
      contentSettings = [
        req.contentType !== undefined ? `"${req.contentType}"` : '"*"',
        `"${req.contentLanguage}"`,
      ];
    } else if (req.contentType !== undefined) {
      contentSettings = [`"${req.contentType}"`];
    }

    const requestStatements: UnparserTokenGroup[] = [];

    if (req.query !== undefined) {
      requestStatements.push(
        Tok.withSpaces('query', this.visitObjectLiteralNode(req.query))
      );
    }
    if (req.headers !== undefined) {
      requestStatements.push(
        Tok.withSpaces('headers', this.visitObjectLiteralNode(req.headers))
      );
    }
    if (req.body !== undefined) {
      if (req.body.kind === 'ObjectLiteral') {
        requestStatements.push(
          Tok.withSpaces('body', this.visitObjectLiteralNode(req.body))
        );
      } else {
        requestStatements.push(
          Tok.withSpaces('body', '=', this.visit(req.body))
        );
      }
    }

    // TODO lol
    return [
      ...security,
      ...Tok.withSpaces(
        'request',
        ...contentSettings,
        Tok.statementBlock(...requestStatements)
      ),
    ];
  }

  visitHttpResponseHandlerNode(
    hand: HttpResponseHandlerNode
  ): UnparserTokenGroup {
    const status =
      hand.statusCode !== undefined ? hand.statusCode.toString() : undefined;
    let contentFilters: string[] = [];
    if (hand.contentLanguage !== undefined) {
      contentFilters = [
        hand.contentType !== undefined ? `"${hand.contentType}"` : '"*"',
        `"${hand.contentLanguage}"`,
      ];
    } else if (hand.contentType !== undefined) {
      contentFilters = [`"${hand.contentType}"`];
    }

    return Tok.withSpaces(
      'response',
      status,
      ...contentFilters,
      Tok.statementBlock(...hand.statements.map(s => this.visit(s)))
    );
  }

  visitHttpCallStatementNode(http: HttpCallStatementNode): UnparserTokenGroup {
    let service: UnparserToken | string = Tok.if(
      'httpCallDefaultService',
      Tok.atom('default')
    );
    if (http.serviceId !== undefined) {
      service = `"${http.serviceId}"`;
    }

    const statements: UnparserTokenGroup[] = [];
    if (http.request !== undefined) {
      statements.push(this.visitHttpRequestNode(http.request));
    }
    statements.push(...http.responseHandlers.map(h => this.visit(h)));

    return Tok.withSpaces(
      'http',
      http.method,
      service,
      `"${http.url}"`,
      Tok.statementBlock(...statements)
    );
  }

  visitMapDefinitionNode(map: MapDefinitionNode): UnparserTokenGroup {
    this.currentContext = 'map';

    return Tok.withSpaces(
      'map',
      map.name,
      Tok.statementBlock(...map.statements.map(s => this.visit(s)))
    );
  }

  visitMapHeaderNode(header: MapHeaderNode): UnparserTokenGroup {
    const profile =
      ProfileId.fromParameters({
        scope: header.profile.scope,
        name: header.profile.name,
      }).withoutVersion +
      '@' +
      VersionRange.fromParameters({
        major: header.profile.version.major,
        minor: header.profile.version.minor,
      }).toString();

    const result = [
      ...Tok.withSpaces('profile', '=', `"${profile}"`),
      Tok.newline(),
      ...Tok.withSpaces('provider', '=', `"${header.provider}"`),
      Tok.newline(),
    ];
    if (header.variant !== undefined) {
      result.push(
        ...Tok.withSpaces('variant', '=', `"${header.variant}"`),
        Tok.newline()
      );
    }

    return result;
  }

  visitOperationDefinitionNode(
    operation: OperationDefinitionNode
  ): UnparserTokenGroup {
    this.currentContext = 'operation';

    return Tok.withSpaces(
      'operation',
      operation.name,
      Tok.statementBlock(...operation.statements.map(s => this.visit(s)))
    );
  }

  visitOutcomeStatementNode(outcome: OutcomeStatementNode): UnparserTokenGroup {
    let keyword;
    if (this.currentContext === 'map') {
      if (outcome.isError) {
        keyword = 'map error';
      } else {
        keyword = 'map result';
      }

      if (outcome.terminateFlow) {
        keyword = 'return ' + keyword;
      }
    } else {
      if (outcome.isError) {
        keyword = 'fail';
      } else {
        keyword = 'return';
      }
    }

    return Tok.withSpaces(
      keyword,
      outcome.condition ? this.visit(outcome.condition) : undefined,
      this.visit(outcome.value)
    );
  }

  visitInlineCallNode(call: InlineCallNode): UnparserTokenGroup {
    return this.visitCallHead(call);
  }

  visitMapDocumentNode(document: MapDocumentNode): UnparserTokenGroup {
    return [
      ...this.visit(document.header),
      Tok.newline(),
      ...Tok.addGlue(
        [Tok.newline(), Tok.newline()],
        document.definitions.map(d => this.visit(d))
      ),
    ];
  }
}
