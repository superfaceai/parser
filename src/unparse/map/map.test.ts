import { readFile } from 'fs/promises';
import { join as joinPath } from 'path';

import { Source } from '../../common';
import { parseMap } from '../../language';
import { DEFAULT_MAP_UNPARSER_CONFIGURATION } from './configuration';
import { MapUnparser } from './visitor';

async function readFixture(name: string): Promise<string> {
  return readFile(joinPath('fixtures', 'unparse', name), { encoding: 'utf-8' });
}

describe('map unparser', () => {
  it.each([
    ['social-media.publish-post.pinterest.suma'],
    ['full-example.suma'],
  ])('map %s', async name => {
    const map = await readFixture(name);
    const ast = parseMap(new Source(map));

    const unparser = new MapUnparser(DEFAULT_MAP_UNPARSER_CONFIGURATION);
    const result = unparser.unparse(ast);

    // expect(result).toBe(map); // use for debugging, should be only whitespace and implicit default parameter differences (as per style settings)
    expect(result).toMatchSnapshot();
  });
});
