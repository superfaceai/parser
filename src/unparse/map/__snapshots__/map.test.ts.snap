// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map unparser map full-example.suma 1`] = `
"profile = \\"example/test@1.2\\"
provider = \\"giver\\"
variant = \\"special\\"

map Foo {
  foo = 1

  foo.bar = true

  set {
    \\"quz.a\\" = \\"hello\\"
    foo.\\"bar-man\\".qux = 1 + 2
  }

  set if (condition === true) {
    fuz = foo.bar
  }

  set {
    foo = call Fop(arg1 = \`String goes \${here}\`, arg2 = \\"hello\\" + \\" world\\")
    foo = call foreach (x of array) Fop(arg = x)
    foo = call Fop(arg = x) if (x % 2 === 0)
    foo = call foreach (x of array) Fop(arg = x) if (x % 2 === 0)
  }

  call Fop(arg2 = 1) {
    foo = 1

    map result {
      a = 1
      b = 2
    }

    map result if (cond) \\"what\\"
  }

  call foreach (x of array) Fop(arg = 1) {
    return map result {
      x = \\"hello\\"
    }

    return map result if (cond === false) {
      banana = true
    }
  }

  call Fop(arg = 1) if (x % 2 === 0) {
    map error \\"ERROR\\"

    map error if (outcome.error) outcome.error
  }

  call foreach (x of array) Fop(arg = 1) if (x % 2 === 0) {
    return map error {
      e = \\"rror\\"
    }

    return map error if (outcome.error) {
      e = outcome.error
      b = \\"ye\\"
    }
  }

  map result true

  http GET \\"/api/v1/resource/{key}/{ another.key}\\" {

  }

  http PUT \\"/api/{url}\\" {
    request {

    }

    response {

    }
  }

  http POST \\"service-name\\" \\"/api/thing\\" {
    security \\"bearer\\"
    request \\"application/json\\" \\"en-US\\" {
      query {
        foo = 1
      }
    }

    response 200 {

    }

    response \\"application/json\\" {

    }

    response \\"*\\" \\"en-US\\" {

    }
  }

  http GET \\"/api\\" {
    request \\"text/plain\\" {
      headers {
        \\"my-header.here\\" = \\"hello\\"
      }

      body = {
        key: \\"value\\",
        thing: true
      }
    }

    response \\"*\\" \\"en-US\\" {
      a = 1
    }

    response 400 \\"*\\" \\"en-US\\" {
      map result if (body.ok) 7

      map error if (!body.ok) 7
    }
  }

  http GET \\"/api\\" {
    request \\"*\\" \\"en-US\\" {
      query {
        foo = 1
        bar = 2
      }

      headers {
        oh = \\"no\\"
      }

      body = [1, 2, 3]
    }

    response 300 \\"application/json\\" \\"en-US\\" {
      return map result 1

      return map error {
        a = \\"hello\\"
      }
    }
  }
}

operation Fop {
  foo = 1

  foo.bar = true

  set {
    \\"quz.a\\" = \\"hello\\"
    foo.\\"bar-man\\".qux = 1 + 2
  }

  set if (condition === true) {
    fuz = foo.bar
  }

  set {
    foo = call Fop(arg1 = \`String goes \${here}\`, arg2 = \\"hello\\" + \\" world\\")
    foo = call foreach (x of array) Fop(arg = x)
    foo = call Fop(arg = x) if (x % 2 === 0)
    foo = call foreach (x of array) Fop(arg = x) if (x % 2 === 0)
  }

  call Fop(arg2 = 1) {
    foo = 1

    return {
      a = 1
      b = 2
    }

    return if (cond) \\"what\\"
  }

  call foreach (x of array) Fop(arg = 1) {
    return {
      x = \\"hello\\"
    }

    return if (cond === false) {
      banana = true
    }
  }

  call Fop(arg = 1) if (x % 2 === 0) {
    fail \\"ERROR\\"

    fail if (outcome.error) outcome.error
  }

  call foreach (x of array) Fop(arg = 1) if (x % 2 === 0) {
    fail {
      e = \\"rror\\"
    }

    fail if (outcome.error) {
      e = outcome.error
      b = \\"ye\\"
    }
  }

  return true

  http GET \\"/api/v1/resource/{key}/{ another.key}\\" {

  }

  http PUT \\"/api/{url}\\" {
    request {

    }

    response {

    }
  }

  http POST \\"service-name\\" \\"/api/thing\\" {
    security \\"bearer\\"
    request \\"application/json\\" \\"en-US\\" {
      query {
        foo = 1
      }
    }

    response 200 {

    }

    response \\"application/json\\" {

    }

    response \\"*\\" \\"en-US\\" {

    }
  }

  http GET \\"/api\\" {
    request \\"text/plain\\" {
      headers {
        \\"my-header.here\\" = \\"hello\\"
      }

      body = {
        key: \\"value\\",
        thing: true
      }
    }

    response \\"*\\" \\"en-US\\" {
      a = 1
    }

    response 400 \\"*\\" \\"en-US\\" {
      return if (body.ok) 7

      fail if (!body.ok) 7
    }
  }

  http GET \\"/api\\" {
    request \\"*\\" \\"en-US\\" {
      query {
        foo = 1
        bar = 2
      }

      headers {
        oh = \\"no\\"
      }

      body = [1, 2, 3]
    }

    response 300 \\"application/json\\" \\"en-US\\" {
      return 1

      fail {
        a = \\"hello\\"
      }
    }
  }
}"
`;

exports[`map unparser map social-media.publish-post.pinterest.suma 1`] = `
"profile = \\"social-media/publish-post@1.1\\"
provider = \\"pinterest\\"

map PublishPost {
  return map error if (!Array.isArray(input.media) || input.media.length !== 1) {
    title = \\"Invalid media\\"
    detail = \\"You must provide exactly one media item for Pinterest.\\"
  }

  origMedia = input.media

  set {
    media = origMedia[0]
    mediaSource = {}
  }

  set if (media.contents) {
    mediaSource = {
      source_type: 'image_base64',
      data: Buffer.isBuffer(media.contents) ? media.contents.toString('base64') : media.contents,
      content_type: undefined,
    }
  }

  set if (mediaSource.data && mediaSource.data.startsWith('/9j')) {
    mediaSource.content_type = \\"image/jpeg\\"
  }

  set if (mediaSource.data && mediaSource.data.startsWith('iVB')) {
    mediaSource.content_type = \\"image/png\\"
  }

  set if (media.url) {
    mediaSource = {
      source_type: 'image_url',
      url: media.url,
    }
  }

  return map error if (!mediaSource.url && !mediaSource.content_type) {
    title = \\"Unknown or unsupported media contents type\\"
    detail = \\"Only PNG or JPG images are supported.\\"
  }

  http POST \\"/v5/pins\\" {
    request {
      headers {
        authorization = \\"Bearer \\" + parameters.accessToken
      }

      body {
        board_id = input.profileId
        link = input.link
        description = input.text
        alt_text = media.altText
        media_source = mediaSource
      }
    }

    response 201 \\"application/json\\" {
      map result {
        postId = body.id
        url = \`https://www.pinterest.com/pin/\\\\\${body.id}/\`
      }
    }

    response 400 \\"application/json\\" {
      map error {
        title = \\"Bad request\\"
        detail = body.message
        original = body
      }
    }

    response 401 \\"application/json\\" {
      map error {
        title = \\"Unauthenticated\\"
        detail = body.message
        original = body
      }
    }

    response 403 \\"application/json\\" {
      map error {
        title = \\"Forbidden\\"
        detail = body.message
        original = body
      }
    }

    response 404 \\"application/json\\" {
      map error {
        title = \\"Not found\\"
        detail = body.message
        original = body
      }
    }

    response 429 \\"application/json\\" {
      map error {
        title = \\"Too Many Requests\\"
        detail = body.message
        original = body
      }
    }

    response {
      map error {
        title = \\"Unexpected error\\"
        detail = body.message
        original = body
      }
    }
  }
}"
`;
